sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +312,95 @@ def predict(self, X):
         """
         check_is_fitted(self)
 
-        X = self._validate_data(X, accept_sparse="csr", reset=False)
-        return self.classes_[
-            pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
-        ]
+        if len(np.unique(self.class_priors_, return_counts=True)[0]) == 1:
+            X = self._validate_data(X, accept_sparse="csr", reset=False)
+            return self.classes_[
+                pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
+            ]
+        else:
+            d = self._decision_function(X)
+            return self.classes_.take(d.argmax(1))
+
+    def _decision_function(self, X):
+        check_is_fitted(self, "centroids_")
+
+        X = self._validate_data(X, reset=False, accept_sparse="csr")
+
+        # check if this works with sparse matrix formats later
+        discriminant_score = np.empty(
+            (X.shape[0], self.classes_.size), dtype=np.float64
+        )
+
+        for cur_class in range(self.classes_.size):
+            Xdist = X - self.centroids_[cur_class, :]
+            Xdist_norm = np.square(Xdist / self.within_class_std_)
+            # Hastie et al. (2009), p. 652, Eq. (18.2)
+            discriminant_score[:, cur_class] = np.squeeze(
+                -np.sum(Xdist_norm, axis=1)
+                + 2.0 * np.log(self.class_priors_[cur_class])
+            )
+
+        return discriminant_score
+
+    def decision_function(self, X):
+        """Apply decision function to an array of samples.
+
+        The decision function is equal (up to a constant factor) to the
+        log-posterior of the model, i.e. `log p(y = k | x)`.
+        However, the decision function uses the argmax definition instead
+        of argmin definition as in the predict() function, which are just
+        -1 factor of each other.
+        The estimation has been implemented according to
+        Hastie et al. (2009), p. 652 equation (18.2)
```

Comment:
I would not go in details here but only refer to ESLII and the equation.

Reply:
@adrinjalali wants to leave these in. do you have second thoughts on this? i also prefer leaving it in since for bizarre reason, [Wiki's article](https://en.wikipedia.org/wiki/Nearest_centroid_classifier) on it is rather lacking and the term "decision function" isn't explicitly used in the [original article](https://pmc.ncbi.nlm.nih.gov/articles/PMC124443/pdf/pq1002006567.pdf).

Reply:
The decision function was moved to DiscriminantAnalysisPredictionMixin class, so I'm adding the following quick note to NearestCentroid's __decision_function_: `# return log posterior, see eq (18.2) p. 652 of the ESL.`